url026791d51b_dhleong_SeriesTracker_v6_tgz-pJ8-stv6_SeriesTrackerJ8/src/stv6/handlers/PluginHandler.java

```diff
@@ -0,0 +1,128 @@
+package stv6.handlers;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStreamReader;
+
+import stv6.Profile;
+import stv6.http.request.Request;
+import stv6.http.request.RequestHandler;
+import stv6.http.request.Response;
+
+public class PluginHandler implements RequestHandler {
+	public static final String HANDLED_PAGE = "plugins";
+
+	@Override
+	public String getHandledPage() {
+		return HANDLED_PAGE;
+	}
+
+	@Override
+	public boolean handle(Request r, Response resp) {
+		String pluginName = r.path.substring(r.path.indexOf('/', 1)+1);
+		String args = null;
+		int argsPos = pluginName.indexOf('?');
+		if (argsPos > -1) {
+			args = pluginName.substring(argsPos+1);
+			pluginName = pluginName.substring(0, argsPos);
+		}
+		File exe = Profile.getInstance().getPluginExe();
+//		if (exe == null || !exe.exists()) 
+//			return false; 
+			
+		File plugin = new File("plugins/" + pluginName);
+		if (plugin == null || !plugin.exists()) 
+			return false;
+		
+		try {			
+			String[] cmdarray;
+			if ((exe == null || !exe.exists()) && plugin.canExecute()) {
+				// no exe, but the plugin itself can be executed
+				cmdarray = (args != null) ? 
+						new String[] {plugin.getAbsolutePath(), args} :
+						new String[] {plugin.getAbsolutePath()};
+			} else {
+				cmdarray = (args != null) ? 
+					new String[] {exe.getAbsolutePath(), plugin.getAbsolutePath(), args} :
+					new String[] {exe.getAbsolutePath(), plugin.getAbsolutePath()};
+			}
+//			System.out.print("Executing: ");
+//			for (String c : cmdarray)
+//				System.out.print(c +"\n");
+//			System.out.println();
+			Process p = Runtime.getRuntime().exec(cmdarray);
+			
+			// read the results right into the response
+			p.getInputStream().mark(2000); // should be way more than enough
+			BufferedReader stdInput = new BufferedReader(
+					new InputStreamReader(p.getInputStream()));
+			String ln = null;
+			
+			// first, read any headers
+			String contentType = null;
+			int contentLength = 0; // we require Content-length for binary!
+			int read = 0;
+			while ((ln = stdInput.readLine()) != null) {
+//				System.out.println(ln);
+				if (ln.length() == 0) {
+					read += 2;
+					break;
+				}
+				if (ln.startsWith("Content-type")) 
+					contentType = ln.substring(13); // get the type
+				if (ln.startsWith("Content-length")) {
+					try {
+						contentLength = Integer.parseInt(ln.substring(15));
+					} catch (NumberFormatException e) {
+						contentLength = 0;
+					}
+				}
+				read += ln.length() + 1; // +1 for newline
+			}
+			
+			if (contentType != null) {
+				resp.setContentType(contentType);
+				
+				if (!(contentType.startsWith("application") || 
+						contentType.startsWith("text")) &&
+						contentLength > 0) {
+					// binary data
+					byte[] data  = new byte[contentLength];
+					p.getInputStream().reset();
+					p.getInputStream().skip(read);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Check the number of bytes that were skipped when you called `skip()`. Handle cases when fewer than the expected number of bytes were skipped. See the following diffs on GitHub that make this code change: [diff1](https://github.com/apache/tomcat80/commit/facf983d8a7952e8d6b03204cb6610fc5f981dc6#diff-e58d7c084e48bd454de5beb71712f72d243bd0287ae2096d7b771f3a2dae6fb0R174), [diff2](https://github.com/apache/hadoop/commit/45620eee687b1093c812873a99833dab068e43d2#diff-f1ae2a4b3016a9deb983f2562d1e71372baf960cf5122d474a37a00f0307edb2R45)

Learn more: [InputStream.skip in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#skip-long-)